Save iframe resources to disk

szewai opened the pull request:
#### 53dd791f052cd7ee77e99c492a0b480166fb12b0
<pre>
Save iframe resources to disk
<a href="https://bugs.webkit.org/show_bug.cgi?id=262567">https://bugs.webkit.org/show_bug.cgi?id=262567</a>
rdar://116420990

Reviewed by Ryosuke Niwa.

This patch adds support for saving iframe resource to disk. To achive that:
1. each iframe main resource is saved as a single file, where file name is decided based on frame ID.
2. iframe main resources and subresources are stored in [main_resources_file_name]_files directory as other subresources
of main frame.
3. subresource urls in iframe main resources are changed to relative file paths.
4. iframe urls in main frame resource are changed to relative file paths.

Test: WebArchive.SaveResourcesSubframe

* Source/WebCore/editing/MarkupAccumulator.cpp:
(WebCore::MarkupAccumulator::appendStartTag):
(WebCore::MarkupAccumulator::frameForAttributeReplacement):
(WebCore::MarkupAccumulator::replaceAttributeIfNecessary): replace srcdoc attribute with src attribute whose value is
relative file path.
(WebCore::MarkupAccumulator::appendURLAttributeIfNecessary): if src or srcdoc attribute does not exist, create a new
src attribute with url set to relative file path.
* Source/WebCore/editing/MarkupAccumulator.h:
* Source/WebCore/loader/archive/cf/LegacyWebArchive.cpp:
(WebCore::LegacyWebArchive::create):
* Source/WebCore/loader/archive/cf/LegacyWebArchive.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/CreateWebArchive.mm:

Canonical link: <a href="https://commits.webkit.org/268858@main">https://commits.webkit.org/268858@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/98bfaab7624f73f39c005c36f503c5052b84b724

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/20739 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/21153 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/21810 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/22631 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/19348 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/20976 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/24401 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/21336 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/20664 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/20961 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/20765 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/18019 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/23485 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/17931 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/18825 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/25144 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/19009 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/19003 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/23039 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/19598 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/16652 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/18827 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5011 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/23159 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/19416 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/editing/MarkupAccumulator.cpp:
```diff
@@ -542,6 +553,46 @@ QualifiedName MarkupAccumulator::xmlAttributeSerialization(const Attribute& attr
+    if (!is<HTMLIFrameElement>(currentElement))
+        return nullptr;
+
+    auto& iframeElement = downcast<HTMLIFrameElement>(*currentElement);
```
How about frameset?
https://developer.mozilla.org/en-US/docs/Web/HTML/Element/frameset

szewai replied:
The link says not recommended, and I think we usually don't add support for to-be-deprecated interfaces. But maybe we do want to ensure the existence of frameset does not fail the saving; I will check on that.

rniwa replied:
I don't think whether a feature is deprecated or not has much bearing in terms of supporting it in the serialization. The fact of matter is this feature is used in wild and it's going to be confusing for users if the feature didn't work with frameset.

szewai replied:
That's reasonable; I will follow up on `<frame>` support.

rniwa commented on Source/WebCore/editing/MarkupAccumulator.cpp:
```diff
@@ -542,6 +553,46 @@ QualifiedName MarkupAccumulator::xmlAttributeSerialization(const Attribute& attr
+        return attribute;
+
+    auto frame = frameForAttributeReplacement(element);
+    if (!frame || !frame->loader().documentLoader()->response().url().isAboutSrcDoc())
```
How about data URL? It seems like we can skip this step as well.

szewai replied:
Thanks for bringing this up. Data URL and blob URL are on my to-do list and I am going to look at them next.

rniwa commented on Source/WebCore/loader/archive/cf/LegacyWebArchive.cpp:
```diff
@@ -535,17 +535,25 @@ RefPtr<LegacyWebArchive> LegacyWebArchive::create(const String& markupString, Lo
+                auto subframeMainResource = subframeArchive->mainResource();
+                auto subframeMainResourceURL = subframeMainResource ? subframeMainResource->url() : URL { };
+                if (!subframeMainResourceURL.isNull()) {
+                    if (subframeMainResourceURL.isAboutSrcDoc() || subframeMainResourceURL.isAboutBlank())
```
Does this work for data URL & blob URL?

rniwa commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/CreateWebArchive.mm:
```diff
@@ -211,6 +211,106 @@
+</script>
+<iframe onload="onFrameLoaded();" id="iframe1_id"></iframe>
+<iframe onload="onFrameLoaded();" src="iframe.html"></iframe>
+<iframe onload="onFrameLoaded();" srcdoc="<p>iframe3</p><img src='image.png'>"></iframe>
```
We should probably also test data URL & blob URL as well as `frameset`.

szewai committed:
Save iframe resources to disk
https://bugs.webkit.org/show_bug.cgi?id=262567
rdar://116420990

Reviewed by Ryosuke Niwa.

This patch adds support for saving iframe resource to disk. To achive that:
1. each iframe main resource is saved as a single file, where file name is decided based on frame ID.
2. iframe main resources and subresources are stored in [main_resources_file_name]_files directory as other subresources
of main frame.
3. subresource urls in iframe main resources are changed to relative file paths.
4. iframe urls in main frame resource are changed to relative file paths.

Test: WebArchive.SaveResourcesSubframe

* Source/WebCore/editing/MarkupAccumulator.cpp:
(WebCore::MarkupAccumulator::appendStartTag):
(WebCore::MarkupAccumulator::frameForAttributeReplacement):
(WebCore::MarkupAccumulator::replaceAttributeIfNecessary): replace srcdoc attribute with src attribute whose value is
relative file path.
(WebCore::MarkupAccumulator::appendURLAttributeIfNecessary): if src or srcdoc attribute does not exist, create a new
src attribute with url set to relative file path.
* Source/WebCore/editing/MarkupAccumulator.h:
* Source/WebCore/loader/archive/cf/LegacyWebArchive.cpp:
(WebCore::LegacyWebArchive::create):
* Source/WebCore/loader/archive/cf/LegacyWebArchive.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/CreateWebArchive.mm:

Canonical link: https://commits.webkit.org/268858@main
